Twister: fix terminology

nashif opened the pull request:
An attempt to overhaul twister naming conventions and get it to closer to the structure of tests we have now. This shall avoid confusion and eliminates legacy naming of various classes and how we process results.

Still work in progress.

nashif committed:
twister: TestSuite -> TestRunner

nashif committed:
twister: TestCase -> TestApplication

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: testcase.yaml -> tests.yaml

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: use test scenarios where it applies

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: capture ztest suitename

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: testcase -> testsuite

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: use Test Scenario termonilogy

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: add testcase class

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: add testcase class

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: remove old release data

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: fix parsing of single test results

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: enhance reporting structure

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: fix ztest testcase

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: disable footprint report (wip)

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: remove temporary debugging

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: do not indent json output

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: misc fixes

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

hakehuang commented on scripts/twister:
```diff
@@ -556,12 +559,12 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
         "-S", "--enable-slow", action="store_true",
         help="Execute time-consuming test cases that have been marked "
-             "as 'slow' in testcase.yaml. Normally these are only built.")
+             "as 'slow' in tests.yaml. Normally these are only built.")
```
I think this should be testcase.yaml not tests.yaml.

hakehuang commented on scripts/twister:
```diff
@@ -556,12 +559,12 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
         "-K", "--force-platform", action="store_true",
         help="""Force testing on selected platforms,
-        even if they are excluded in the test configuration (testcase.yaml)"""
+        even if they are excluded in the test configuration (tests.yaml)"""
```
I think this should be testcase.yaml not tests.yaml.

hakehuang reviewed (approved):
I guess there are two miss replacements from 'testcase.yaml` to 1tests.yaml`, comments in line. others looks good to me

PerMac commented:
@hakehuang I believe these are not missed replacements. One of the reasons behind this PR was issues with the naming convention, in particular, "test case" was problematic as it was used to name both test cases defined in .c and test suites (collections of test cases).

PerMac commented:
@nashif At the first glance, the reports look fine. I started reviewing the code a while ago but decided to wait until you finish the pr. I will also ask you if some clean ups in the commits can be done (e.g. there are lines that are changed multiple times in multiple commits, hence reviewing on the commit base is almost impossible). 
I think the structure of `instance` objects in the code should be modified: now you have `self.instance.scenario.suite.(instances/name)` i.e. `suite` is sitting inside `scenario`. `suite` also has `scenarios` field, so the relation looks a bit convoluted. In fact, when debugging in PyCharm it looks like an infinite loop: `self.instance.scenario.suite.instances[0].scenario.suite.instances ...`
From my understanding `suite` is at a higher level of tests' organization than `scenario`.

hakehuang commented:
> @hakehuang I believe these are not missed replacements. One of the reasons behind this PR was issues with the naming convention, in particular, "test case" was problematic as it was used to name both test cases defined in .c and test suites (collections of test cases).

@PerMac , we only have testcase.yaml not test.yaml.

PerMac commented:
@hakehuang For now, yes. But the point is to switch testcase.yaml -> tests.yaml. This PR adds support for this: 
https://github.com/zephyrproject-rtos/zephyr/pull/32541/commits/c0405b18d674a67c947c97519a4cec8ae4a58e91
So the default will be tests.yaml but testcase.yaml will also be checked for legacy reason

nashif commented:
> I will also ask you if some clean ups in the commits can be done (e.g. there are lines that are changed multiple times in multiple commits, hence reviewing on the commit base is almost impossible).

at some point many of those will be squashed, I suggest to look at the change as a whole right now, since this change is about one thing only  and not trying to introduce new features..

nashif committed:
twister: always run with CMAKE_EXPORT_COMPILE_COMMANDS

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: generate build coverage

Generate build coverage statistics. We look at all compiled files per
test and count all *.c files in the git repo and calculate the
statistics based on that.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: adapt tests

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
actions: clang: install gitpython

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: support multiple toolchain variants

Partially implemented, still need to deal with toolchain filters in test
scenarios.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
